docs/src/manual/containers.md

```diff
@@ -485,3 +485,100 @@ JuMP.Containers.SparseAxisArray{Int64, 2, Tuple{Int64, Int64}} with 2 entries:
   [1, 2]  =  3
   [1, 4]  =  5
 ```
+
+## Keyword indexing
+
+Opt-in to the experimental keyword indexing syntax by passing
+`enable_keyword_indexing = true` to the [`Containers.@container`](@ref) macro,
+or call [`enable_container_keyword_indexing`](@ref).
+
+Keyword indexing lets you uses the named indices of [`Containers.DenseAxisArray`](@ref)
+and [`Containers.SparseAxisArray`](@ref) as keyword arguments when indexing a
+container:
+
+```jldoctest container_kwarg_indexing
+julia> Containers.@container(
+           x[i=2:4, j=1:3],
+           i + j,
+           enable_keyword_indexing = true,
+       )
+2-dimensional DenseAxisArray{Int64,2,...} with index sets:
+    Dimension 1, 2:4
+    Dimension 2, Base.OneTo(3)
+And data, a 3×3 Matrix{Int64}:
+ 3  4  5
+ 4  5  6
+ 5  6  7
+
+julia> x[i = 2, j = 1]
+3
+
+julia> Containers.@container(
+           y[i=2:4, j=1:3; i > j],
+           i + j,
+           enable_keyword_indexing = true,
+       )
+JuMP.Containers.SparseAxisArray{Int64, 2, Tuple{Int64, Int64}} with 6 entries:
+  [2, 1]  =  3
+  [3, 1]  =  4
+  [3, 2]  =  5
+  [4, 1]  =  5
+  [4, 2]  =  6
+  [4, 3]  =  7
+
+julia> y[i = 2:4, j = 1]
+JuMP.Containers.SparseAxisArray{Int64, 1, Tuple{Int64}} with 3 entries:
+  [2]  =  3
+  [3]  =  4
+  [4]  =  5
+
+julia> using JuMP
+
+julia> model = Model();
+
+julia> @variable(model, x[i = 2:3])
+1-dimensional DenseAxisArray{VariableRef,1,...} with index sets:
+    Dimension 1, 2:3
+And data, a 2-element Vector{VariableRef}:
+ x[2]
+ x[3]
+
+julia> x[i = 2]
+ERROR: Keyword indexing is disabled. To enable, pass `enable_keyword_indexing = true` to the `Containers.@container` macro, or call `JuMP.enable_container_keyword_indexing(model, true)` before calling any JuMP macros like `@variable`.
+Stacktrace:
+[...]
+
+julia> model = Model();
+
+julia> enable_container_keyword_indexing(model, true)
+
+julia> @variable(model, x[i = 2:3])
+1-dimensional DenseAxisArray{VariableRef,1,...} with index sets:
+    Dimension 1, 2:3
+And data, a 2-element Vector{VariableRef}:
+ x[2]
+ x[3]
+
+julia> x[i = 2]
+x[2]
+```
+
+The main reason this syntax is currently opt-in is that it does not work for
+`Array`:
+
+```jldoctest container_kwarg_indexing
+julia> Containers.@container(
+           z[i=1:4, j=1:3],
+           i + j,
+           enable_keyword_indexing = true,
```

Comment:
I don't know if we can do this without throwing misleading errors if you create an array using `@variable` after calling `enable_container_keyword_indexing(model, true)`. 

Few people actually use `Containers.@container`, and setting this option requires reading the docstring and the limitations are documented.

Reply:
Could we make it so that if `enable_container_keyword_indexing(model, true)` is called or if `enable_keyword_indexing = true` and there are keywords then we create a `DenseAxisArray` instead of an `Array`. Then we never have to throw any error.

Reply:
Because I don't think keyword indexing should take priority over returning an Array.

I still think we're over thinking this. People don't get confused that `x[1:2]` returns an `Array` but `x[S]` where `S=1:2` returns a `DenseAxisArray`. 

I'd still prefer that this was turned on by default, and we deal with any confusion with better documentation _once we have feedback that it is indeed an issue_.

Reply:
> Because I don't think keyword indexing should take priority over returning an Array.

If you pass the keyword argument to a macro and then you return an `Array`, it doesn't make any sense.
The user explictly said he wanted keyword indexing so you can't return a container that does not.
For the model level flag, I would also return a `DenseAxisArray` when names indices are used at creation

Reply:
This now triggers an error.